packages/babel-parser/src/plugins/typescript/index.ts

```diff
@@ -593,21 +601,37 @@ export default (superClass: ClassWithMixin<typeof Parser, IJSXParserMixin>) =>
       return this.finishNode(node, "TSImportType");
     }
 
-    tsParseEntityName(allowReservedWords: boolean = true): N.TsEntityName {
-      let entity: N.TsEntityName = this.parseIdentifier(allowReservedWords);
+    tsParseEntityName(flags: tsParseEntityNameFlags): N.TsEntityName {
+      let entity: N.TsEntityName;
+      if (
+        flags & tsParseEntityNameFlags.ALLOW_THIS_EXPRESSION &&
+        this.match(tt._this)
+      ) {
+        const node = this.startNode<N.ThisExpression>();
+        this.next();
+        entity = this.finishNode(node, "ThisExpression");
+      } else {
+        entity = this.parseIdentifier(
```

Comment:
If we don't pass `ALLOW_THIS_EXPRESSION`, `this` would still be parsed but as an identifier right? If so, I'd prefer to call it `SPECIAL_THIS`, `THIS_AS_EXPRESSION`, or (the inverse) `THIS_AS_IDENTIFIER`